spyder/config/user.py

```diff
@@ -5,369 +5,498 @@
+        TODO:
+    raw_mode: bool
+        TODO:
+    remove_obsolete: bool
+        TODO:
+
+    Notes
+    -----
+    The 'get' and 'set' arguments number and type differ from the overriden
+    methods. 'defaults' is an attribute and not a method.
     """
     DEFAULT_SECTION_NAME = 'main'
-    def __init__(self, name, defaults=None, load=True, version=None,
-                 subfolder=None, backup=False, raw_mode=False,
-                 remove_obsolete=False):
-        DefaultsConfig.__init__(self, name, subfolder)
-        self.raw = 1 if raw_mode else 0
-        if (version is not None and
-                re.match(r'^(\d+).(\d+).(\d+)$', version) is None):
-            raise ValueError("Version number %r is incorrect - must be in X.Y.Z format" % version)
-        if isinstance(defaults, dict):
-            defaults = [ (self.DEFAULT_SECTION_NAME, defaults) ]
-        self.defaults = defaults
-        if defaults is not None:
-            self.reset_to_defaults(save=False)
-        fname = self.filename()
-        if backup:
-            try:
-                shutil.copyfile(fname, "%s.bak" % fname)
-            except IOError:
-                pass
+
+    def __init__(self, name, path, defaults=None, load=True, version=None,
+                 backup=False, raw_mode=False, remove_obsolete=False):
+        """UserConfig class, based on ConfigParser."""
+        super(UserConfig, self).__init__(name=name, path=path)
+
+        self._load = load
+        self._version = self._check_version(version)
+        self._backup = backup
+        self._raw = 1 if raw_mode else 0
+        self._remove_obsolete = remove_obsolete
+        self._module_source_path = get_module_source_path('spyder')
+
+        self._defaults_folder = 'defaults'
+        self._backup_folder = 'backups'
+        self._backup_suffix = '.bak'
+        self._defaults_name_prefix = 'defaults'
+
+        # This attribute is overriding a method from cp.ConfigParser
+        self.defaults = self._check_defaults(defaults)
+
         if load:
-            # If config file already exists, it overrides Default options:
-            self.load_from_ini()
-            old_ver = self.get_version(version)
-            _major = lambda _t: _t[:_t.find('.')]
-            _minor = lambda _t: _t[:_t.rfind('.')]
+            # If config file already exists, it overrides Default options
+            previous_fpath = self.get_previous_config_fpath()
+            self._load_from_ini(previous_fpath)
+            old_version = self.get_version(version)
+            self._old_version = old_version
+
             # Save new defaults
-            self._save_new_defaults(defaults, version, subfolder)
+            self._save_new_defaults(defaults)
+
             # Updating defaults only if major/minor version is different
-            if _minor(version) != _minor(old_ver):
+            if (self._get_minor_version(version=version)
+                    != self._get_minor_version(version=old_version)):

```

Comment:
Let's remove the `version` kwarg here because I don't think it's necessary. So this should be

    if (self._get_minor_version(version)
            != self._get_minor_version(old_version):

Reply:
Agreed